jdbc/src/main/java/nextstep/jdbc/JdbcTemplate.java

```diff
@@ -1,9 +1,86 @@
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import javax.sql.DataSource;
+import nextstep.jdbc.exception.DataAccessException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 public class JdbcTemplate {
 
-    private static final Logger log = LoggerFactory.getLogger(JdbcTemplate.class);
+    private static final Logger LOG = LoggerFactory.getLogger(JdbcTemplate.class);
+
+    private final DataSource dataSource;
+
+    public JdbcTemplate(DataSource dataSource) {
+        this.dataSource = dataSource;
+    }
+
+    public int update(String sql, Object... args) {
+        return execute(PreparedStatement::executeUpdate, sql, args);
+    }
+
+    public <T> List<T> query(String sql, RowMapper<T> rowMapper, Object... args) {
+        return execute(preparedStatement -> executeQueryAndMappingList(preparedStatement, rowMapper), sql, args);
+    }
+
+    private <T> List<T> executeQueryAndMappingList(PreparedStatement preparedStatement, RowMapper<T> rowMapper) {
+        try (ResultSet resultSet = preparedStatement.executeQuery()){
+            List<T> results = new ArrayList<>();
+
+            while (resultSet.next()) {
+                results.add(rowMapper.map(resultSet));
+            }
+
+            return results;
+        } catch (SQLException e){
+            throw new DataAccessException(e.getMessage());
+        }
+    }
+
+    public <T> Optional<T> queryForObject(String sql, RowMapper<T> rowMapper, Object... args) {
+        return execute(preparedStatement -> executeQueryAndMapping(preparedStatement, rowMapper), sql, args);
+    }
+
+    private <T> Optional<T> executeQueryAndMapping(PreparedStatement preparedStatement, RowMapper<T> rowMapper) {
+        try (ResultSet resultSet = preparedStatement.executeQuery()){
+            if (resultSet.next()) {
+                return Optional.of(rowMapper.map(resultSet));
+            }
+
+            return Optional.empty();
+        } catch (SQLException e) {
+            throw new DataAccessException(e.getMessage());
+        }
+    }
+
+    private <T> T execute(PreparedStatementCallback<T> preparedStatementCallback, String sql, Object... args) {
+        LOG.debug("query : {}", sql);
+
+        try (
+            Connection connection = dataSource.getConnection();
+            PreparedStatement preparedStatement = createStatement(connection, sql, args)
+        ) {
+            return preparedStatementCallback.call(preparedStatement);
+        } catch (SQLException e) {
+            LOG.error(e.getMessage(), e);
+            throw new DataAccessException(e.getMessage());
+        }
+    }
+
+    private PreparedStatement createStatement(Connection connection, String sql, Object[] args) throws SQLException {
+        PreparedStatement preparedStatement = connection.prepareStatement(sql);
+
+        for (int i = 0; i < args.length; i++) {
+            preparedStatement.setObject(i + 1, args[i]);
+        }
```

Comment:
for문이 한 싸이클 돌 때 마다 `args.length()` 메서드가 호출될 것 같아요 🤔
`args.length()`를 변수로 추출하여 추출한 변수를 사용하거나 혹은 아래와 같이 Stream을 활용해봐도 좋을 것 같아요 😃

```suggestion
        IntStream.range(0, args.length)
                .forEach(it -> preparedStatement.setObject(it + 1, args[it]));
```

Reply:
이 부분에 대해서 피드백이 올거같았는데!! 👍  

> (다른 이야기를 하시다가 예시로) 
"예를 들어 for문 루프를 돌릴 때 기준 변수에 `list.size()`를 사용하는게 성능상 이슈가 발생하겠지만, 그 이슈가 현재 하드웨어 스펙으로 충분히 메꿀 수 있는 수준이고, 개발자끼리 코드를 읽을 때 더 도움이 된다면 충분한 트레이드 오프가 되지 않을까요?"

레벨1~2 사이에 포비께서 예시로 해주신 말씀이에요!! 그래서 변수로 뺄까... 하다가 for문에 포함시켜봤어요!!
`forEach`의 경우 for문과 달리 루프중에 예외가 발생할 경우에도 멈추지 않고 모든 동작을 수행한다고 알고있어서 사용하기 무섭더라구요 😨 ..

`개발자끼리 코드를 읽을 때 더 도움이 된다면 충분한 트레이드 오프가 되지 않을까요?`를 생각한다면 로키와 같이 읽는 코드를 생각했을 때 변수로 추출하는게 가장 이상적인 방법인거 같네요!! 바로 적용하겠습니다!! 👍

Reply:
흠 변수로 빼는 정도는 사실 그렇게 어려운 부분이 아니고 코드의 가독성을 떨어뜨린다고 생각하지는 않는데 🤔  
인용하신 피드백 부분의 글을 읽어보니 어느정도 공감이 가네요 😄 
그래도 변수로 추출하는 정도라면 저의 의견과 현구막이 이전에 받은 피드백을 둘 다 충족할 수 있을 것 같네요 👍 

> forEach가 루프중에 예외가 발생할 경우에도 멈추지 않고 모든 동작을 수행한다 

는 사실을 잘 모르고 있었는데 하나 배워가네요 😉